interception-details.md

```diff
@@ -0,0 +1,215 @@
+1. `appHistory.current` updates. `appHistory.current.finished` is `false`.
+1. `currentchange` fires on `window.appHistory`.
+1. `appHistory.current` fires `navigateto`.
+1. Any now-unreachable `AppHistoryEntry` instances fire `dispose`.
+1. The `console.log()` outputs `"#foo"`
+
+Asynchronously but basically immediately:
+
+1. The URL bar updates.
+
+After the promise settles in one microtask:
+
+1. `appHistory.current.finished` changes to `true`.
+1. `appHistory.current` fires `finish`.
+1. `navigatefinish` is fired on `appHistory`.
+
+### No interception
+
+```js
+appHistory.addEventListener("navigate", e => {
+  // Do nothing.
+});
+
+location.hash = "#foo";
+console.log(location.hash);
+```
+
+This is the same as the previous case.
+
+### Cancelation
+
+```js
+appHistory.addEventListener("navigate", e => {
+  e.preventDefault();
+});
+
+location.hash = "#foo";
+console.log(location.hash);
+```
+
+Synchronously:
+
+1. `navigate` fires on `window.appHistory`. The event gets canceled.
+1. The `console.log()` outputs `""` (or whatever the old hash was).
+
+## Same-origin `respondWith()`, cross-document
+
+When performing a same-origin navigations, `respondWith()` can be used to take over the navigation and convert it to a same-document navigation, even if it would normally be a cross-document navigation.
+
+### Delayed success
+
+```js
+appHistory.addEventListener("navigate", e => {
+  e.respondWith(new Promise(r => setTimeout(r, 10_000)));
+});
+
+location.href = "/foo";
+console.log(location.href);
+```
+
+Synchronously:
+
+1. `navigate` fires on `window.appHistory`.
+1. `appHistory.current` fires `navigatefrom`.
+1. `location.href` updates.
+1. `appHistory.current` updates. `appHistory.current.finished` is `false`.
+1. `currentchange` fires on `window.appHistory`.
+1. `appHistory.current` fires `navigateto`.
+1. Any now-unreachable `AppHistoryEntry` instances fire `dispose`.
+1. The `console.log()` outputs `"/foo"`.
+
+Asynchronously but basically immediately:
+
+1. The URL bar updates to show `/foo`.
+1. Any loading spinner UI starts.
+
+After the promise fulfills in ten seconds:
+
+1. `appHistory.current.finished` changes to `true`.
+1. `appHistory.current` fires `finish`.
+1. `navigatefinish` is fired on `appHistory`.
+1. Any loading spinner UI stops.
+
+### Delayed failure
+
+```js
+appHistory.addEventListener("navigate", e => {
+  e.respondWith(new Promise((r, reject) => setTimeout(() => reject(new Error("bad")), 10_000)));
+});
+
+location.href = "/foo";
+console.log(location.href);
+
+setTimeout(() => {
+  console.log(location.href);
+}, 10_000);
+```
+
+Synchronously:
+
+1. `navigate` fires on `window.appHistory`.
+1. `appHistory.current` fires `navigatefrom`.
+1. `location.href` updates.
+1. `appHistory.current` updates. `appHistory.current.finished` is `false`.
+1. `currentchange` fires on `window.appHistory`.
+1. `appHistory.current` fires `navigateto`.
+1. Any now-unreachable `AppHistoryEntry` instances fire `dispose`.
```

Comment:
for delayed failures, what happens to these "now-unreachable" Entries? should they become "reachable" again since the navigation was cancelled and `appHistory.current` was rolled back to its previous value?

Reply:
Great question.

I'm inclined to say that they're gone forever? Splicing a future history back into existence seems pretty weird. But in general the introduction of fallible single-page navigations raises a lot of questions, hmm...

Reply:
I'll open a dedicated issue about this and then point to it from here.